File: src/p0xxx/p315.py
from heapq import heappop, heappush
from typing import List


class Solution:
    def countSmaller(self, nums: List[int]) -> List[int]:
        counts = [0] * len(nums)
        indices = list(range(len(nums)))
        indices2 = indices.copy()

        def mergesort(start, stop, indices, indices2):
            count = stop - start
            if count <= 1:
                return

                # print('mergesort', start, stop, nums[start:stop], counts[start:stop])
            mid = (start + stop) // 2

            mergesort(start, mid, indices2, indices)
            mergesort(mid, stop, indices2, indices)

            # print('merge idx', left, right)
            # print('merge num', [nums[s] for s in left], [nums[s] for s in right])

            smaller_count = 0
            i, i_stop = start, mid
            j, j_stop = mid, stop
            k, k_stop = start, stop

            while i < i_stop and j < j_stop:
                a = nums[indices2[i]]
                b = nums[indices2[j]]

                if a <= b:
                    indices[k] = indices2[i]
                    counts[indices2[i]] += smaller_count
                    k += 1
                    i += 1
                else:
                    indices[k] = indices2[j]
                    smaller_count += 1
                    k += 1
                    j += 1

            while i < i_stop:
                indices[k] = indices2[i]
                counts[indices2[i]] += smaller_count
                k += 1
                i += 1

            while j < j_stop:
                indices[k] = indices2[j]
                k += 1
                j += 1

        mergesort(0, len(nums), indices, indices2)

        return counts

    def countSmaller5(self, nums: List[int]) -> List[int]:
        # Binary Index Tree
        size = len(nums)
        tree = [0] * (size + 1)

        sorted_compact_nums = sorted(set(nums))
        table = {s: i for i, s in enumerate(sorted_compact_nums)}

        def update(idx: int, val: int):
            idx += 1
            while idx <= size:
                tree[idx] += val
                idx += idx & (-idx)

        def sum_to(idx: int):
            partial = 0
            idx += 1
            while idx > 0:
                partial += tree[idx]
                idx -= idx & (-idx)

            return partial

        print('compact', sorted_compact_nums)
        print('tree', tree)

        res = []
        for i in range(len(nums))[::-1]:
            idx = table[nums[i]]
            res.append(sum_to(idx - 1))
            update(idx, 1)

            print('compact', sorted_compact_nums)
            print('tree', tree)

        res.reverse()
        return res

    def countSmaller4(self, nums: List[int]) -> List[int]:
        counts = [0] * len(nums)

        def mergesort(start, stop):
            count = stop - start
            if count <= 1:
                return [start] if count > 0 else []

            # print('mergesort', start, stop, nums[start:stop], counts[start:stop])
            mid = (start + stop) // 2

            left = mergesort(start, mid)
            right = mergesort(mid, stop)

            # print('merge idx', left, right)
            # print('merge num', [nums[s] for s in left], [nums[s] for s in right])

            smaller_count = 0
            indices = []
            i, i_stop = 0, len(left)
            j, j_stop = 0, len(right)

            while i < i_stop and j < j_stop:
                a = nums[left[i]]
                b = nums[right[j]]

                if a <= b:
                    indices.append(left[i])
                    counts[left[i]] += smaller_count
                    i += 1
                else:
                    indices.append(right[j])
                    smaller_count += 1
                    j += 1

            while i < i_stop:
                indices.append(left[i])
                counts[left[i]] += smaller_count
                i += 1

            while j < j_stop:
                indices.append(right[j])
                j += 1

            return indices

        mergesort(0, len(nums))

        return counts

    def countSmaller3(self, nums: List[int]) -> List[int]:
        counts = [0] * len(nums)

        indices = list(range(len(nums)))
        indices.sort(key=lambda s: nums[s])
        incomplete = [True] * len(nums)
        print(indices)

        for idx in indices:
            for i in range(idx):
                if incomplete[i]:
                    counts[i] += 1

            incomplete[idx] = False

        return counts

    def countSmaller2(self, nums: List[int]) -> List[int]:
        counts = [0] * len(nums)
        stack = []  # increasing from head to tail
        maxheap = []

        for i in range(len(nums))[::-1]:
            number = nums[i]

            while stack and stack[-1] < number:
                heappush(maxheap, -stack.pop())

            while maxheap and -maxheap[0] >= number:
                stack.append(-heappop(maxheap))

            # print(i, number, stack, maxheap)
            counts[i] = len(maxheap)
            heappush(maxheap, -number)

        return counts

File: src/p0xxx/p315.rs
impl Solution {
    pub fn count_smaller(nums: Vec<i32>) -> Vec<i32> {
        use ::std::collections::HashMap;

        // sort indices by values
        let mut indices: Vec<usize> = (0..nums.len()).collect();
        indices.sort_by_key(|&s| nums[s]);

        // form a virtual sorted nums by indices which does not need to exist
        // sorted_unique_nums = sorted(set(nums))

        // build mapping number -> their index
        let mut table = HashMap::<i32, usize>::with_capacity(nums.len());
        indices.into_iter().enumerate().for_each(|(i, s)| {
            table.entry(nums[s]).or_insert(i);
        });

        // binary index tree
        // prefix sums
        let mut tree = vec![0i32; nums.len() + 1];
        let mut counts = vec![0i32; nums.len()];

        #[inline(always)]
        fn sum_before(tree: &mut [i32], mut idx: usize) -> i32 {
            let mut sum = 0;
            while idx > 0 {
                sum += tree[idx];
                let s = idx as isize;
                idx -= (s & (-s)) as usize;
            }
            sum
        }

        #[inline(always)]
        fn update(tree: &mut [i32], mut idx: usize, val: i32) {
            idx += 1;
            while idx < tree.len() {
                tree[idx] += val;
                let s = idx as isize;
                idx += (s & (-s)) as usize;
            }
        }

        // println!("{:?}", table);

        for i in (0..nums.len()).rev() {
            let number = nums[i];
            let idx = *table.get(&number).unwrap();

            // sum values between 0 up to idx (exclusive)
            counts[i] = sum_before(&mut tree, idx);
            // update values between 0 up to idx (exclusive)
            update(&mut tree, idx, 1);

            // println!("{}: num={} idx={}, tree={:?}", i, number, idx, tree);
        }

        counts
    }

    pub fn count_smaller2(nums: Vec<i32>) -> Vec<i32> {
        let mut counts = vec![0i32; nums.len()];
        let mut indices = (0..nums.len()).collect::<Vec<usize>>();
        let mut indices2 = indices.clone();

        fn mergesort(
            nums: &[i32],
            counts: &mut [i32],
            indices: &mut [usize],
            indices2: &mut [usize],
            start: usize,
            stop: usize,
        ) {
            let count = stop - start;
            if count <= 1 {
                return;
            }

            let mid = (start + stop) / 2;

            mergesort(nums, counts, indices2, indices, start, mid);
            mergesort(nums, counts, indices2, indices, mid, stop);

            let (mut i, i_stop) = (start, mid);
            let (mut j, j_stop) = (mid, stop);
            let mut k = start;
            let mut smaller_count = 0;

            while i < i_stop && j < j_stop {
                let a = nums[indices2[i]];
                let b = nums[indices2[j]];

                if a <= b {
                    indices[k] = indices2[i];
                    counts[indices2[i]] += smaller_count;
                    i += 1;
                    k += 1;
                } else {
                    indices[k] = indices2[j];
                    smaller_count += 1;
                    j += 1;
                    k += 1;
                }
            }

            for i in i..i_stop {
                indices[k] = indices2[i];
                counts[indices2[i]] += smaller_count;
                k += 1;
            }

            for j in j..j_stop {
                indices[k] = indices2[j];
                k += 1;
            }
        }

        mergesort(
            &nums,
            &mut counts,
            &mut indices,
            &mut indices2,
            0,
            nums.len(),
        );

        counts
    }
}

fn main() {
    Solution::count_smaller(vec![5, 2, 6, 1]);

    Solution::count_smaller(vec![-1]);

    Solution::count_smaller(vec![-1, -1]);
    println!("Hello World");
}
